Current list of speakers view (fixes #2212)

matakuka opened the pull request:
The idea is creating two views.
1) A view within the normal OpenSlides interface (/speakers), similar to /agenda or /motions. The content div of this page may be rescaled to fullscreen, but if not rescaled, it behaves as the normal interface does (navigation headers, sidebars, chat visible, etc.).
2) A view which is always reachable via a different URL (/listofspeakers), similar to /projector, being independent of whatever the user is doing on the normal interface. This view _only_ shows the last, current and next speakers or a message "no current speakers", and no header/footer/navigation tools.

However, I am stuck on the issue, with an error message I cannot decipher. What fails and keeps me from continuing with the issue seems to be something in the connection between the controller (agenda/site.js) and the template (/agenda/item-list-of-speakers.html) . The error message has about 40 lines pointing to obscure points in static/js/openslides-libs.js, beginning with: 
```
"Error: [ng:areq] Argument 'Itemundefined' is not a function, got undefined
http://errors.angularjs.org/1.4.11/ng/areq?p0=Itemundefined&p1=not%20aNaNunction%2C%20got%20undefinedminErr/
```
with no helpful hints on what is wrong; and changing anything in the template (including sending a blank template) doesn't help. (Also, searching for "Itemundefined" returned nothing)

normanjaeckel commented:
See #2212.

emanuelschuetze commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -71,6 +77,9 @@ angular.module('OpenSlidesApp.agenda.site', ['OpenSlidesApp.agenda'])
+            })
+            .state('agenda.item.list-of-speakers', {
+                   url: '/speakers',
             });
```
controller: 'AgendaListOfSpeakersCtrl'

emanuelschuetze commented on openslides/agenda/static/templates/agenda/item-list-of-speakers.html:
```diff
@@ -0,0 +1,43 @@
+  </div>
+</div>
+
+<div ng-controller="AgendaListOfSpeakersCtrl" class="content">
```
without ng-controller

matakuka commented:
Status of this PR:

@emanuelschuetze and I discussed this PR and came to the conclusion, that a second view in a different browser window is not what we want at the moment, but rather only a `<div>` that can be extended to full screen by a `ng-click` event on a button. 
(to discuss: another button on this full screen to return to "normal" screen or just a click on the extended div container?). The outlines of this div should be generic, so it can be reused on different divs throughout openSlides.

The frame for this is already more or less present, the major missing part is the actual filling of the page with information (and some cleaning up).

As the problem (guessing which agenda_item is the current one *on a completely different part of the program*) is similar to #2211, I may put the resulting code from there to a different place (not `/core/projector`) where *this* page can access it as well, so this issue is dependent on the other one.

emanuelschuetze commented:
Travis failed in jshint, see openslides/agenda/static/js/agenda/site.js.

matakuka commented:
ready for testing @emanuelschuetze

emanuelschuetze commented:
Please rebase on top of master. Then I'll do some style improvements.

matakuka commented:
rebased on top of master @emanuelschuetze

emanuelschuetze commented on openslides/core/static/css/app.css:
```diff
@@ -798,6 +797,7 @@ img {
     content: ":";
 }
 
+<<<<<<< 67612401a75ae384c850d65bc3dfb5cb0f811596
```
please remove

normanjaeckel commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -71,6 +77,24 @@ angular.module('OpenSlidesApp.agenda.site', ['OpenSlidesApp.agenda'])
+                   url: '/speakers',
+                   controller: 'ListOfSpeakersViewCtrl',
+                   resolve: {
+                       users: function(User) {
```
Indetion is broken.

normanjaeckel commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -546,6 +570,115 @@ angular.module('OpenSlidesApp.agenda.site', ['OpenSlidesApp.agenda'])
+                function() {
+                    return Projector.lastModified(1);
+                },
+                function(scope) {
```
`scope` argument is not necessary.

normanjaeckel commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -546,6 +570,115 @@ angular.module('OpenSlidesApp.agenda.site', ['OpenSlidesApp.agenda'])
+        };
+
+        //run this on loading of the page, because we want to get the projector even if unchanged
+        $scope.checkoutItem();
```
Is this function automaticly called when the watch is loaded the first time?

normanjaeckel commented on openslides/core/static/css/app.css:
```diff
@@ -832,6 +832,38 @@ img {
 
 #groups-table .optional-show { /* hide optional-show column */
     display: none;
+=======
```
Here are also some merge fragments ...

matakuka commented:
@normanjaeckel ready for review

normanjaeckel commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -71,6 +71,21 @@ angular.module('OpenSlidesApp.agenda.site', ['OpenSlidesApp.agenda'])
+                   url: '/speakers',
+                   controller: 'ListOfSpeakersViewCtrl',
+                   resolve: {
+                       users: function(User) {
```
Idention is broken.

normanjaeckel commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -546,6 +561,52 @@ angular.module('OpenSlidesApp.agenda.site', ['OpenSlidesApp.agenda'])
+            },
+            function() {
+                Projector.find(1).then( function(projector) {
+                $scope.AgendaItem = null;
```
Please indent this block.

normanjaeckel commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -546,6 +561,52 @@ angular.module('OpenSlidesApp.agenda.site', ['OpenSlidesApp.agenda'])
+                Projector.find(1).then( function(projector) {
+                $scope.AgendaItem = null;
+                _.forEach(projector.elements, function(element) {
+                    if (element.name == 'motions/motion') {
```
Suggestion: Use a switch case block. (This is only a suggestion. You can also open a new issue and we add the label clean up for later.)

normanjaeckel commented on openslides/agenda/static/templates/agenda/current-list-of-speakers.html:
```diff
@@ -0,0 +1,46 @@
+    </h2>
+  </div>
+  <div class="content">
+  <!-- Last speakers -->
```
Indention is not correct.

normanjaeckel commented on openslides/core/static/css/app.css:
```diff
@@ -1087,3 +1130,4 @@ tr.selected td {
         width: 150px;
     }
 }
+
```
If there are two newlines at the end of the file, please remove one. If not, not. :-)

normanjaeckel commented:
Some styleing remarks.

matakuka commented:
done. @normanjaeckel

normanjaeckel commented:
Ok.

emanuelschuetze commented on openslides/agenda/static/templates/agenda/current-list-of-speakers.html:
```diff
@@ -0,0 +1,48 @@
+        <translate>Fullscreen</translate>
+      </button>
+      <button os-perms="agenda.can_manage"
+        ng-click="goToListOfSpeakers()"
```
this function is still missing in site.js (go to does not work)

emanuelschuetze commented:
Ok, looks good. Please add the missing goto function and then rebase to one commit. I'll merge it today.

matakuka commented:
@emanuelschuetze 
rebased, merged, added the function and changed layout of the  button (goto list). 

TODO (low priority): Could not find a good icon for fullscreen

emanuelschuetze commented:
Ok looks good. Ready for merge now...

> TODO (low priority): Could not find a good icon for fullscreen

@matakuka Please make a new issue for that.
